rfc/rfc-80/rfc-80.md

```diff
@@ -0,0 +1,169 @@
+3. After sorting, split the data into column families.  
+4. Write the segmented data into the log file of the corresponding column family.  
+
+#### Common API interface
+After the table columns are clustered, the writing process includes the process of sorting and splitting the data compared to the original bucket bucketing. A new append interface needs to be introduced to support column families.  
+Introduce ColumnFamilyAppendHandle extend AppendHandle to implement column family writing.
+
+![append-handle](append-handle.png)
+
+### Reading data
+#### ColumnFamilyReader and RowReader
+![row-reader](row-reader.png)
+
+Hudi internal row reader reading steps:  
+1. Hudi organizes files by column families to be read.
+2. Introduce familyReader to merge and read each column family's own baseFile and logfile to achieve column family-level data reading.  
+    * Since log files are written after being sorted by primary key, familyReader merges its own baseFile and logFile by primary key using sortMerge.
+    * familyReader supports upstream incoming column pruning to reduce IO overhead when reading data.  
+    * During the merging process, if the user specifies the precombie field for the column family, the merging strategy will be selected based on the precombie field. This logic reuses Hudi's own precombine logic and does not need to be modified.    
+3. Row reader merges the data read by multiple familyReaders according to the primary key.  
+
+Since the data read by each familyReader is sorted by the primary key, the row reader merges the data read by each familyReader in the form of sortMergeJoin and returns the complete data.  
+
+The entire reading process involves a large amount of data merging, but because the data itself is sorted, the memory consumption of the entire merging process is very low and the merging is fast. Compared with Hudi's native merging method, the memory pressure and the merging time are significantly reduced.
+
+#### Engine reads pseudo process
+![process-read](process-read.png)
+
+1) The engine itself delivers the data files that need to be scanned to executor/woker/taskmanger.  
+2) executor/worker/taskmanger calls Hudi’s rowReader interface and passes in column clipping and filter conditions to rowReader.  
+3) The Hudi kernel completes the data reading of rowReader and returns complete data. The data format is Avro.  
+4) The engine gets the Avro format data and needs to convert it into the data format it needs. For example, spark needs to be converted into unsaferow, hetu into block, flink into row, and hive into arrayWritable.
+
+### Column family level compaction
+Extend Hudi's compaction schedule module to merge each column family's own base file and log file:
+
+![family-compaction](family-compaction.png)
+
+### Full compaction
+Extend Hudi's compaction schedule module to merge and update all column families in the entire table.  
```

Comment:
Full compaction could be optional.
If there is no requirement to read across column family, there is no need to do full compaction, right?

Reply:
Yes, it could. But if you don't need to read across column families, may be it's better to use separate tables for each column family instead of one table with many column families.

Reply:
I think full compaction could be optional, only column family level compaction is required.
Different users might need different columns, some might need columns comes from multiple column family, some might need columns from only one column family.
It's better to allow users to choose whether enable full compaction or not.

Besides, after full compaction, projection on the reader side is less efficient because the projection could only be based on the full parquet file with all complete fields instead of based on column family names.

Reply:
It is useful to build a wide table for AI feature table.

Reply:
Agree with you, make it optional is the best decision.